File: src/bmsx_cpp/machine/devices/audio/event_latch.cpp
#include "machine/devices/audio/event_latch.h"

#include "machine/bus/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/irq/controller.h"
#include "machine/memory/memory.h"

namespace bmsx {

ApuEventLatch::ApuEventLatch(Memory& memory, IrqController& irq)
	: m_memory(memory)
	, m_irq(irq) {}

void ApuEventLatch::reset() {
	m_eventSequence = 0;
	m_memory.writeValue(IO_APU_EVENT_KIND, valueNumber(static_cast<double>(APU_EVENT_NONE)));
	m_memory.writeValue(IO_APU_EVENT_SLOT, valueNumber(0.0));
	m_memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, valueNumber(0.0));
	m_memory.writeValue(IO_APU_EVENT_SEQ, valueNumber(0.0));
}

ApuEventLatchState ApuEventLatch::captureState() const {
	ApuEventLatchState state;
	state.eventSequence = m_eventSequence;
	state.eventKind = m_memory.readIoU32(IO_APU_EVENT_KIND);
	state.eventSlot = m_memory.readIoU32(IO_APU_EVENT_SLOT);
	state.eventSourceAddr = m_memory.readIoU32(IO_APU_EVENT_SOURCE_ADDR);
	return state;
}

void ApuEventLatch::restoreState(const ApuEventLatchState& state) {
	m_eventSequence = state.eventSequence;
	m_memory.writeValue(IO_APU_EVENT_KIND, valueNumber(static_cast<double>(state.eventKind)));
	m_memory.writeValue(IO_APU_EVENT_SLOT, valueNumber(static_cast<double>(state.eventSlot)));
	m_memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, valueNumber(static_cast<double>(state.eventSourceAddr)));
	m_memory.writeValue(IO_APU_EVENT_SEQ, valueNumber(static_cast<double>(m_eventSequence)));
}

void ApuEventLatch::emit(u32 kind, ApuAudioSlot slot, u32 sourceAddr) {
	m_eventSequence += 1u;
	m_memory.writeValue(IO_APU_EVENT_KIND, valueNumber(static_cast<double>(kind)));
	m_memory.writeValue(IO_APU_EVENT_SLOT, valueNumber(static_cast<double>(slot)));
	m_memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, valueNumber(static_cast<double>(sourceAddr)));
	m_memory.writeValue(IO_APU_EVENT_SEQ, valueNumber(static_cast<double>(m_eventSequence)));
	m_irq.raise(IRQ_APU);
}

} // namespace bmsx

File: src/bmsx/machine/devices/audio/event_latch.ts
import {
	APU_EVENT_NONE,
	type ApuAudioSlot,
} from './contracts';
import {
	IO_APU_EVENT_KIND,
	IO_APU_EVENT_SEQ,
	IO_APU_EVENT_SLOT,
	IO_APU_EVENT_SOURCE_ADDR,
	IRQ_APU,
} from '../../bus/io';
import type { Memory } from '../../memory/memory';
import type { IrqController } from '../irq/controller';

export type ApuEventLatchState = {
	eventSequence: number;
	eventKind: number;
	eventSlot: number;
	eventSourceAddr: number;
};

export class ApuEventLatch {
	private eventSequence = 0;

	public constructor(
		private readonly memory: Memory,
		private readonly irq: IrqController,
	) {}

	public reset(): void {
		this.eventSequence = 0;
		this.memory.writeValue(IO_APU_EVENT_KIND, APU_EVENT_NONE);
		this.memory.writeValue(IO_APU_EVENT_SLOT, 0);
		this.memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, 0);
		this.memory.writeValue(IO_APU_EVENT_SEQ, 0);
	}

	public captureState(): ApuEventLatchState {
		return {
			eventSequence: this.eventSequence,
			eventKind: this.memory.readIoU32(IO_APU_EVENT_KIND),
			eventSlot: this.memory.readIoU32(IO_APU_EVENT_SLOT),
			eventSourceAddr: this.memory.readIoU32(IO_APU_EVENT_SOURCE_ADDR),
		};
	}

	public restoreState(state: ApuEventLatchState): void {
		this.eventSequence = state.eventSequence >>> 0;
		this.memory.writeValue(IO_APU_EVENT_KIND, state.eventKind);
		this.memory.writeValue(IO_APU_EVENT_SLOT, state.eventSlot);
		this.memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, state.eventSourceAddr);
		this.memory.writeValue(IO_APU_EVENT_SEQ, this.eventSequence);
	}

	public emit(kind: number, slot: ApuAudioSlot, sourceAddr: number): void {
		this.eventSequence = (this.eventSequence + 1) >>> 0;
		this.memory.writeValue(IO_APU_EVENT_KIND, kind);
		this.memory.writeValue(IO_APU_EVENT_SLOT, slot);
		this.memory.writeValue(IO_APU_EVENT_SOURCE_ADDR, sourceAddr);
		this.memory.writeValue(IO_APU_EVENT_SEQ, this.eventSequence);
		this.irq.raise(IRQ_APU);
	}
}
